File: bin/ruby-cli/bin/merge-markdown.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '../lib/archive/file_merger_base'
require 'pathname'

# Merge markdown files with their references into a single file
class MergeMarkdown < FileMergerBase
  def script_title
    'Merge Markdown Files'
  end

  def script_description
    'Merges markdown files into a single document.'
  end

  def script_arguments
    '<input_file> [output_file]'
  end

  def file_extension
    'md'
  end

  def add_custom_options(opts)
    super
    opts.on('-p', '--preserve-structure', 'Preserve directory structure in headers') do
      @options[:preserve_structure] = true
    end
  end

  def validate!
    super

    @input_file = File.expand_path(@args[0])
    unless File.exist?(@input_file)
      log_error("Input file not found: #{@input_file}")
      exit 1
    end

    unless File.extname(@input_file).downcase == '.md'
      log_error('Input file must be a markdown (.md) file')
      exit 1
    end

    # Set default options
    @options[:preserve_structure] ||= false

    @base_dir = File.dirname(@input_file)
    @processed_files = Set.new
    @file_order = [] # Preserve discovery order
    @file_contents = {}
  end

  def merge_files(files)
    log_progress('Creating merged file...')

    File.open(@output_file, 'w') do |output|
      # Write header
      output.puts '# Merged Documentation'
      output.puts
      output.puts "Generated on: #{Time.now.strftime('%Y-%m-%d %H:%M:%S')}"
      output.puts "Source file: #{relative_path(@input_file)}"
      output.puts "Files merged: #{files.size}"
      output.puts
      output.puts '---'
      output.puts

      # Write files in discovery order
      files.each do |file_path|
        write_file_section(output, file_path)
      end
    end

    log_file_created(@output_file)
    log_success("Merged #{files.size} files into #{relative_path(@output_file)}")
  end

  def collect_files
    log_progress('Scanning for referenced files...')
    scan_file(@args[0])
    @file_order
  end

  private

  def scan_file(file_path)
    return if @processed_files.include?(file_path)

    log_debug("Scanning: #{file_path}")
    @processed_files.add(file_path)
    @file_order << file_path # Preserve discovery order

    unless File.exist?(file_path)
      log_warning("Referenced file not found: #{file_path}")
      return
    end

    content = File.read(file_path)
    @file_contents[file_path] = content

    return unless @options[:recursive]

    # Find markdown file references
    references = find_markdown_references(content, File.dirname(file_path))
    references.each { |ref| scan_file(ref) }
  end

  def find_markdown_references(content, base_dir)
    references = []

    # Match markdown links: [text](path.md) or [text](./path.md) or [text](../path.md)
    content.scan(/\[([^\]]*)\]\(([^)]+\.md)\)/i) do |_text, path|
      # Resolve relative paths
      full_path = if path.start_with?('/')
                    path
                  else
                    File.expand_path(File.join(base_dir, path))
                  end

      references << full_path if File.exist?(full_path)
    end

    # Match file includes: <!-- include: path.md --> or similar
    content.scan(/<!--\s*include:\s*([^>]+\.md)\s*-->/i) do |path|
      full_path = if path.start_with?('/')
                    path
                  else
                    File.expand_path(File.join(base_dir, path))
                  end

      references << full_path if File.exist?(full_path)
    end

    references.uniq
  end

  def write_file_section(output, file_path)
    relative = relative_path(file_path)

    # Create section header
    header = if @options[:preserve_structure]
               "# #{relative}"
             else
               "# #{File.basename(file_path, '.md').gsub(/[-_]/, ' ').split.map(&:capitalize).join(' ')}"
             end

    output.puts header
    output.puts
    output.puts "**Source:** `#{relative}`"
    output.puts

    # Write content, but skip the first H1 header if it exists to avoid duplication
    content = @file_contents[file_path]
    lines = content.lines

    # Skip first line if it's an H1 header
    lines = lines[1..-1] if lines.first&.start_with?('# ')

    output.puts lines.join
    output.puts
    output.puts '---'
    output.puts
  end

  def relative_path(file_path)
    Pathname.new(file_path).relative_path_from(Pathname.new(@base_dir)).to_s
  end

  def show_examples
    puts 'Examples:'
    puts "  #{script_name} README.md"
    puts "  #{script_name} docs/main.md merged_docs.md"
    puts "  #{script_name} --dry-run --preserve-structure project.md"
    puts "  #{script_name} --no-recursive simple.md"
  end
end

# Execute the script
MergeMarkdown.execute if __FILE__ == $0

File: bin/bun-cli/src/scripts/files/merge-markdown.ts
import { Script } from "../../core/decorators/Script";
import { BaseScript } from "../../core/base/Script";
import { logger } from "../../core/utils/logger";
import { readFile, writeFile } from "node:fs/promises";
import { dirname, join, relative, basename, extname } from "node:path";
import { existsSync } from "node:fs";
import * as readline from "node:readline/promises";

/**
 * Merge markdown files with their references into a single file
 */
@Script({
  name: "merge-markdown",
  description: "Merges markdown files into a single document",
  emoji: "📚",
  arguments: "<input_file> [output_file]",
  examples: [
    "merge-markdown README.md",
    "merge-markdown docs/main.md merged_docs.md",
    "merge-markdown --dry-run --preserve-structure project.md",
    "merge-markdown --no-recursive simple.md",
  ],
  options: [
    {
      flag: "-p, --preserve-structure",
      description: "Preserve directory structure in headers",
    },
    {
      flag: "-r, --recursive",
      description: "Recursively scan referenced markdown files (default: true)",
    },
    {
      flag: "--no-recursive",
      description: "Disable recursive scanning",
    },
    {
      flag: "-o, --overwrite",
      description: "Overwrite output file if it exists",
    },
  ],
})
export class MergeMarkdownScript extends BaseScript {
  private inputFile!: string;
  private outputFile!: string;
  private baseDir!: string;
  private processedFiles = new Set<string>();
  private fileOrder: string[] = [];
  private fileContents = new Map<string, string>();
  private options: {
    preserveStructure: boolean;
    recursive: boolean;
    overwrite: boolean;
  } = {
    preserveStructure: false,
    recursive: true,
    overwrite: false,
  };

  async run(args: string[], options: Record<string, any>): Promise<void> {
    logger.banner("Merge Markdown Files");

    // Parse options
    this.options = {
      preserveStructure: options.preserveStructure || false,
      recursive: options.recursive !== false, // Default to true unless explicitly disabled
      overwrite: options.overwrite || false,
    };

    // Validate arguments
    if (args.length === 0) {
      logger.error("Input file is required");
      logger.info("Usage: merge-markdown <input_file> [output_file]");
      process.exit(1);
    }

    this.inputFile = args[0];
    if (!existsSync(this.inputFile)) {
      logger.error(`Input file not found: ${this.inputFile}`);
      process.exit(1);
    }

    if (extname(this.inputFile).toLowerCase() !== ".md") {
      logger.error("Input file must be a markdown (.md) file");
      process.exit(1);
    }

    // Set output file
    if (args.length > 1) {
      this.outputFile = args[1];
    } else {
      const inputBasename = basename(this.inputFile, ".md");
      this.outputFile = join(dirname(this.inputFile), `${inputBasename}-merged.md`);
    }

    // Ensure output file has .md extension
    if (!this.outputFile.endsWith(".md")) {
      this.outputFile += ".md";
    }

    // Check if output file exists
    if (existsSync(this.outputFile) && !this.options.overwrite) {
      logger.error(`Output file '${this.outputFile}' already exists. Use --overwrite to replace.`);
      process.exit(1);
    }

    this.baseDir = dirname(this.inputFile);

    // Collect files
    logger.progress("Scanning for referenced files...");
    await this.scanFile(this.inputFile);

    if (this.fileOrder.length === 0) {
      logger.error("No files found to merge");
      process.exit(1);
    }

    logger.info(`Found ${this.fileOrder.length} files:`);
    this.fileOrder.forEach((file) => console.log(`  📄 ${this.relativePath(file)}`));

    if (options.dryRun) {
      logger.info("[DRY-RUN] Would create merged file");
      logger.completion("Merge Markdown Files");
      return;
    }

    // Confirm action
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    const answer = await rl.question(
      `\nMerge ${this.fileOrder.length} files into '${this.outputFile}'? (y/N): `
    );
    rl.close();

    if (answer.toLowerCase() !== "y") {
      logger.info("Operation cancelled");
      return;
    }

    // Merge files
    await this.mergeFiles();

    logger.completion("Merge Markdown Files");
  }

  private async scanFile(filePath: string): Promise<void> {
    if (this.processedFiles.has(filePath)) {
      return;
    }

    logger.debug(`Scanning: ${filePath}`, false);
    this.processedFiles.add(filePath);
    this.fileOrder.push(filePath);

    if (!existsSync(filePath)) {
      logger.warning(`Referenced file not found: ${filePath}`);
      return;
    }

    const content = await readFile(filePath, "utf-8");
    this.fileContents.set(filePath, content);

    if (!this.options.recursive) {
      return;
    }

    // Find markdown file references
    const references = this.findMarkdownReferences(content, dirname(filePath));
    for (const ref of references) {
      await this.scanFile(ref);
    }
  }

  private findMarkdownReferences(content: string, baseDir: string): string[] {
    const references: string[] = [];

    // Match markdown links: [text](path.md) or [text](./path.md) or [text](../path.md)
    const linkRegex = /\[([^\]]*)\]\(([^)]+\.md)\)/gi;
    let match;
    while ((match = linkRegex.exec(content)) !== null) {
      const path = match[2];
      const fullPath = path.startsWith("/")
        ? path
        : join(baseDir, path);

      if (existsSync(fullPath)) {
        references.push(fullPath);
      }
    }

    // Match file includes: <!-- include: path.md --> or similar
    const includeRegex = /<!--\s*include:\s*([^>]+\.md)\s*-->/gi;
    while ((match = includeRegex.exec(content)) !== null) {
      const path = match[1].trim();
      const fullPath = path.startsWith("/")
        ? path
        : join(baseDir, path);

      if (existsSync(fullPath)) {
        references.push(fullPath);
      }
    }

    return [...new Set(references)]; // Remove duplicates
  }

  private async mergeFiles(): Promise<void> {
    logger.progress("Creating merged file...");

    const sections: string[] = [];

    // Write header
    sections.push("# Merged Documentation\n");
    sections.push(`Generated on: ${new Date().toLocaleString()}`);
    sections.push(`Source file: ${this.relativePath(this.inputFile)}`);
    sections.push(`Files merged: ${this.fileOrder.length}\n`);
    sections.push("---\n");

    // Write files in discovery order
    for (const filePath of this.fileOrder) {
      sections.push(this.createFileSection(filePath));
    }

    await writeFile(this.outputFile, sections.join("\n"));

    logger.fileCreated(this.outputFile);
    logger.success(`Merged ${this.fileOrder.length} files into ${this.relativePath(this.outputFile)}`);
  }

  private createFileSection(filePath: string): string {
    const relativePath = this.relativePath(filePath);
    const content = this.fileContents.get(filePath) || "";

    // Create section header
    let header: string;
    if (this.options.preserveStructure) {
      header = `# ${relativePath}`;
    } else {
      const name = basename(filePath, ".md")
        .replace(/[-_]/g, " ")
        .split(" ")
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(" ");
      header = `# ${name}`;
    }

    const lines = content.split("\n");

    // Skip first line if it's an H1 header to avoid duplication
    const contentLines = lines[0]?.startsWith("# ") ? lines.slice(1) : lines;

    return `${header}\n\n**Source:** \`${relativePath}\`\n\n${contentLines.join("\n")}\n\n---\n`;
  }

  private relativePath(filePath: string): string {
    return relative(this.baseDir, filePath);
  }
}
